packages/graphql-hooks/src/LocalGraphQLClient.ts

```diff
@@ -72,4 +77,57 @@ function timeoutPromise(delayInMs) {
     setTimeout(resolve, delayInMs)
   })
 }
+
+function isObject(o: unknown): o is object {
+  return o === Object(o)
+}
+
+function collectErrorsFromObject(objectIn: object): {data: object | null, errors: Error[]} {
+  const data: object = {}
+  const errors: Error[] = []
+
+  for (const [key, value] of Object.entries(objectIn)) {
+    if (value instanceof Error) {
+      data[key] = null;
+      errors.push(value)
+    } else if (Array.isArray(value)) {
+      const child = collectErrorsFromArray(value)
+      data[key] = child.data
+      errors.push(...child.errors)
+    } else if (isObject(value)) {
+      const child = collectErrorsFromObject(value)
+      data[key] = child.data
+      errors.push(...child.errors)
+    } else {
+      data[key] = value
+    }
+  }
+
+  return {data, errors}
+}
+
+function collectErrorsFromArray(arrayIn: object[]): {data: (object | null)[], errors: Error[]} {
+  const data: (object | null)[] = Array(arrayIn.length)
+  const errors: Error[] = []
+
+  for (const [idx, entry] of arrayIn.entries()) {
+    if (entry instanceof Error) {
+      data[idx] = null
+      errors.push(entry)
+    } else if (Array.isArray(entry)) {
+      const child = collectErrorsFromArray(entry)
+      data[idx] = child.data
+      errors.push(...child.errors)
+    } else if (isObject(entry)) {
+      const child = collectErrorsFromObject(entry)
+      data[idx] = child.data
+      errors.push(...child.errors)
+    } else {
+      data[idx] = entry
+    }
+  }
+
+  return {data, errors}
+}
```

Comment:
I feel like these two functions could be unified somehow, or at least their if-elseif-else logic be extracted to a 3rd function as it is clearly the same for both. The only real differences are the input type, the data init and how the input is traversed.

Everything else looks good to me. 👍